tests/rest/client/test_relations.py

```diff
@@ -241,6 +241,60 @@ def test_repeated_paginate_relations(self):
         found_event_ids.reverse()
         self.assertEquals(found_event_ids, expected_event_ids)
 
+    def test_pagination_from_sync_messages(self):
+        """Pagination tokens from sync and messages should work."""
+        channel = self._send_relation(RelationTypes.ANNOTATION, "m.reaction", "A")
+        self.assertEquals(200, channel.code, channel.json_body)
+        annotation_id = channel.json_body["event_id"]
+        # Send an event after the relation events.
+        self.helper.send(self.room, body="Latest event", tok=self.user_token)
+
+        # Request sync, but limit the timeline to get a pagination token before
+        # the reaction.
+        filter = urllib.parse.quote_plus(
+            '{"room": {"timeline": {"limit": 1}}}'.encode()
+        )
+        channel = self.make_request(
+            "GET", f"/sync?filter={filter}", access_token=self.user_token
+        )
+        self.assertEquals(200, channel.code, channel.json_body)
+        room_timeline = channel.json_body["rooms"]["join"][self.room]["timeline"]
+        sync_prev_batch = room_timeline["prev_batch"]
+        self.assertIsNotNone(sync_prev_batch)
+        # Ensure the event above is not in the batch.
+        self.assertNotIn(
+            annotation_id, [ev["event_id"] for ev in room_timeline["events"]]
+        )
+
+        channel = self.make_request(
+            "GET",
+            f"/rooms/{self.room}/messages?dir=b&limit=1",
+            access_token=self.user_token,
+        )
+        self.assertEquals(200, channel.code, channel.json_body)
+        messages_end = channel.json_body["end"]
+        self.assertIsNotNone(messages_end)
+        # Ensure the event above is not in the chunk.
+        self.assertNotIn(
+            annotation_id, [ev["event_id"] for ev in channel.json_body["chunk"]]
+        )
+
+        # Request the relations with the token from sync/messages -- this is a
+        # tiny bit silly since how do we know the parent ID? Just assume we had
+        # it from an earlier sync and the above are gappy.
```

Comment:
This is a struggle; likely PEBKAC but I suppose it would help me to understand a bit more what's going on here anyway.

the token from sync/messages → from each of /sync and /messages ?

```suggestion
        # Request the relations with the token from each of /sync and /messages -- this is a
        # tiny bit silly since how do we know the parent ID? Just assume we had
        # it from an earlier sync and the above are gappy.
```

Parent ID? Is that the ID of the event which the reaction was sent for?

> and the above are gappy.

I'm struggling to follow that.

Reply:
"Parent ID" is a term used throughout this file, yes. It is the event that the relation is sent against.

I'm trying to say that we're filling in information that the client might be missing. I'll try to expand it.

Reply:
I rewrote a bunch of this in 62cba899b9b63ef6bf53ef7fb6ee54a7294aedd2, hopefully that makes it clearer.